### PR TITLE
[STORM-1052] TridentKafkaState uses new Kafka Producer API

### PR DESCRIPTION
**Pull request by vesense**

### CODE REVIEW

**Comment by vesense** [comment]
@HeartSaVioR  Can you please review this. Thanks.

File: external/storm-kafka/src/test/storm/kafka/TridentKafkaTopology.java [hunk 1]
```diff
@@ -88,11 +88,11 @@ public static void main(String[] args) throws Exception {
...
     private  static Config getConfig(String brokerConnectionString) {
         Config conf = new Config();
-        Map config = new HashMap();
         Properties props = new Properties();
-        props.put("metadata.broker.list", brokerConnectionString);
-        props.put("request.required.acks", "1");
-        props.put("serializer.class", "kafka.serializer.StringEncoder");
+        props.put("bootstrap.servers", brokerConnectionString);
+        props.put("acks", "1");
+        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
```
Comment (harshach) [inline external/storm-kafka/src/test/storm/kafka/TridentKafkaTopology.java]: These should be configurable

File: external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java [hunk 2]
```diff
@@ -72,27 +73,27 @@ public void prepare(Map stormConf) {
...
-                }
-            } catch (Exception ex) {
-                String errorMsg = "Could not send message with key = " + mapper.getKeyFromTuple(tuple)
-                        + " to topic = " + topic;
-                LOG.warn(errorMsg, ex);
-                throw new FailedException(errorMsg, ex);
+        for (final TridentTuple tuple : tuples) {
+            final String topic = topicSelector.getTopic(tuple);
+            if(topic != null) {
+                producer.send(new ProducerRecord(topic, mapper.getKeyFromTuple(tuple),
```
Comment (harshach) [inline external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java]: We are moving from sync behavior to async here. Any reason for this? or can't we just call .get() on the send method.

File: external/storm-kafka/src/test/storm/kafka/TridentKafkaTopology.java (hunk 1 above)
Comment (vesense) [inline external/storm-kafka/src/test/storm/kafka/TridentKafkaTopology.java]: Sorry I took so long to respond. I think it is unnecessary for UT.

File: external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java (hunk 2 above)
Comment (vesense) [inline external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java]: Sure, we should just call get() method to send. I will fix this. Thx.

**Comment by vesense** [comment]
@harshach Could you take a look?

File: external/storm-kafka/src/test/storm/kafka/TridentKafkaTopology.java (hunk 1 above)
Comment (zhuoliu) [inline external/storm-kafka/src/test/storm/kafka/TridentKafkaTopology.java]: Agreed, since it is a test, making it not configurable should be OK.

**Comment by harshach** [comment]
@zhuoliu for trident we are sending the tuples as a batch and we would like to know if any of those tuples failed or any other errors occurred rather than wait for the callback. I think using get here is fine.

File: external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java [hunk 3]
```diff
@@ -83,8 +81,8 @@ public void updateState(List<TridentTuple> tuples, TridentCollector collector) {
                 topic = topicSelector.getTopic(tuple);
 
                 if(topic != null) {
-                    producer.send(new KeyedMessage(topic, mapper.getKeyFromTuple(tuple),
-                            mapper.getMessageFromTuple(tuple)));
+                    producer.send(new ProducerRecord(topic, mapper.getKeyFromTuple(tuple),
```
Comment (harshach) [inline external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java]: we should be checking for exceptions in the response returned by the producer

**Comment by zhuoliu** [comment]
Thanks @harshach for the explanation! That makes good sense.

File: external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java (hunk 3 above)
Comment (vesense) [inline external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaState.java]: Good catch. I will fix it. Thx.

**Comment by zhuoliu** [comment]
Hi @vesense The DisruptorQueue test bug has been fixed in another JIRA. Could you upmerge and trigger the CI again? This pull request looks good to me, I am +1 to merge it in as long as tests pass.

**Comment by revans2** [comment]
I just upmerged and ran the tests on this manually.  Everything looks good for me. I am +1 on merging this in @zhuoliu @harshach @HeartSaVioR do either of you have any more comments you would like to see addressed?

**Comment by zhuoliu** [comment]
+1

**Pull request merged**